enumeratum-core/src/main/scala/enumeratum/Enum.scala

```diff
@@ -85,6 +85,13 @@ trait Enum[A <: EnumEntry] {
     */
   def withNameOption(name: String): Option[A] = namesToValuesMap.get(name)
 
+  /**
+    * Returns an [[Right[A]]] for a given name, or a [[Left[String]]] if the name does not match any of the values'
+    * .entryName values.
+    */
+  def withNameEither(name: String): Either[String, A] =
+    namesToValuesMap.get(name).toRight(buildNotFoundMessage(name))
```

Comment:
I've been thinking about this this some more, and I now think that instead of allocating a string for Left, it might be better (more typesafe, fewer allocations) if we returned an object (e.g. `case object NoSuchMember`).

Reply:
It could be interesting, but it would mean that we would lose the error message that we currently get in Left.
WDYT of using a case class that could hold:

* a reference of the name we tried (that is already allocated in memory, since it was the parameter)
* a reference to the enumeration values (already in memory too)

`toString` could be overridden to produce the error message, only when needed.
There would still be allocations for the case class, but it should be cheaper than allocating the string.

Reply:
Indeed, that could work, and the cost of allocating a case class holding 2 references is cheap, so I'm 👍  to going in that direction.

Reply:
What do you think of the case class extending Throwable. Could be of use to some users (`enableWritableStackTrace = false` if you're concerned with allocation)

Reply:
I wouldn’t mind extending `Exception`, but only if there is no stack trace. We could for instance extend `scala.util.control.NoStackTrace`. 

My stance here is that if some one wants a stack trace, they can throw their own exception on left, it will not be part of this API.

Reply:
Indeed, we could even go as far as extending `NoSuchElementException`, what do you think?